client/client.go

```diff
@@ -1266,61 +1260,154 @@ func (c *Client) addAlloc(alloc *structs.Allocation) error {
 	return nil
 }
 
-// setupConsulClient creates a ConsulService
-func (c *Client) setupConsulClient() error {
-	cs, err := consul.NewConsulService(c.config.ConsulConfig, c.logger)
-	c.consulService = cs
-	return err
-}
-
-// syncConsul removes services of tasks which are no longer in running state
-func (c *Client) syncConsul() {
-	sync := time.NewTicker(consulSyncInterval)
-	for {
-		select {
-		case <-sync.C:
-			// Give up pruning services if we can't fingerprint Consul
+// setupConsulSyncer creates Client-mode consul.Syncer which periodically
+// executes callbacks on a fixed interval.
+//
+// TODO(sean@): this could eventually be moved to a priority queue and give
+// each task an interval, but that is not necessary at this time.
+func (c *Client) setupConsulSyncer() error {
+	// The bootstrapFn callback handler is used to periodically poll
+	// Consul to look up the Nomad Servers in Consul.  In the event the
+	// heartbeat deadline has been exceeded and this Client is orphaned
+	// from its servers, periodically poll Consul to reattach this Client
+	// to its cluster and automatically recover from a detached state.
+	bootstrapFn := func() error {
+		now := time.Now()
+		c.heartbeatLock.Lock()
+
+		// If the last heartbeat didn't contain a leader, give the
+		// Nomad server this Agent is talking to one more attempt at
+		// providing a heartbeat that does contain a leader.
+		if atomic.LoadInt32(&c.lastHeartbeatFromQuorum) == 1 && now.Before(c.consulPullHeartbeatDeadline) {
+			c.heartbeatLock.Unlock()
+			// c.logger.Printf("[TRACE] client.consul: heartbeat received, sleeping until %v", c.consulPullHeartbeatDeadline)
+			return nil
+		}
+		c.heartbeatLock.Unlock()
+		c.logger.Printf("[TRACE] client.consul: lost heartbeat with Nomad quorum, falling back to Consul for server list")
 
-			c.configLock.RLock()
-			_, ok := c.configCopy.Node.Attributes["consul.server"]
-			c.configLock.RUnlock()
-			if !ok {
+		consulCatalog := c.consulSyncer.ConsulClient().Catalog()
+		dcs, err := consulCatalog.Datacenters()
+		if err != nil {
+			return fmt.Errorf("client.consul: unable to query Consul datacenters: %v", err)
+		}
+		dcs = dcs[0:lib.MinInt(len(dcs), datacenterQueryLimit)]
```

Comment:
Can we prefer the local DC as in many cases this will be correct. Also maybe not limit the DCs as it may make it not work depending if the servers aren't in the cutoff set